[12.0][IMP] hr_timesheet_sheet: different validation models

alexey-pelykh opened the pull request:
This module adds option to configure how the timesheet sheet validation is done. Reviewer is
configurable to be HR Officer/Manager, Department Manager, Direct Manager, or
Project Manager.


To change who reviews submitted sheets, go to *Configuration > Settings > Timesheet Options*
and configure **Timesheet Sheet Review Policy** accordingly.

alexey-pelykh committed:
[IMP] hr_timesheet_sheet: validation models

alexey-pelykh commented:
@mreficent @astirpe it would be good to hear out your thoughts on this WIP improvement, especially about
```
        if not self.env.user.has_group('hr_timesheet.group_hr_timesheet_user'):
            raise UserError(
                _('Only an HR Officer or Manager can approve sheets.'))
```
being changed to check for `hr.group_hr_user` as error message sounds right while logic isn't.

MiquelRForgeFlow commented:
I think it should be better if you do a separate module named `hr_timesheet_sheet_review_policy` where you add all this `review_policy` related thing, maybe. I would appreciate that very much.

alexey-pelykh commented:
@mreficent good point to think about. What pros/cons you have in mind?

alexey-pelykh commented:
@mreficent @astirpe from code that's right now in master, it seems that anyone can approve anyone (as long as they have access to timesheets), is that by design?

alexey-pelykh commented:
@mreficent @astirpe this one is ready for review, hopefully you've seen my email about call for discussion

MiquelRForgeFlow commented:
If you don't want to separate the refactoring from the Review Policy thing, then at least do in separate commits. I don't see why all your changes should be compacted in just one commit.

alexey-pelykh commented:
@mreficent I found that decoupling that into separate module would be quite an effort - anyhow "default" review policy would have been implemented in `hr_timesheet_sheet`, so I decided that it would be better to implement most useful policies in this module, using extra 15 lines of code, than extract them into a separate one.

Regarding separating the commits - I added few enhancements to make module more extendable, e.g. extracted couple of methods. That's a pure "refactor"/"more hooks" commit - you mean extract such one?

MiquelRForgeFlow commented:
Yeah, one commit for this "refactor/more hooks", other for the addition of review policy, at least. With separate commits, it's easier to review that having all together.

alexey-pelykh commented:
@mreficent same PR or separate one?

MiquelRForgeFlow commented:
I would prefer separate PR, but I do not want to bother too much :S

alexey-pelykh commented:
https://github.com/OCA/timesheet/pull/264

astirpe commented:
I need some time to check this PR and the other. I hope I can do it this week. Thanks!

alexey-pelykh commented:
@astirpe thanks for your time! 
@mreficent this has been updated and rebased

alexey-pelykh commented:
@mreficent @astirpe 🙏 if I can grab some of your precious time

astirpe commented on hr_timesheet_sheet/models/hr_timesheet_sheet.py:
```diff
@@ -164,6 +199,59 @@ def _default_employee(self):
+            field = view.find(".//field[@name='department_id']")
+
+        if field is not None:
+            field.set('invisible', '0')
```
Shouldn't all this black magic be moved to a separate custom module? I see it's made just for the visibility of two fields in the tree view. I think it doesn't worth it.

alexey-pelykh replied:
But it's the only way to make those columns conditionally invisible, and leaving them as is would only pollute the UI

astirpe replied:
Module `hr_timesheet_sheet` should be a base module for being extended (and thanks for splitting the methods to hook methods) but this kind of changes are going a bit too far IMO. I would prefer to keep the module as a basis, although it became already quite complex compared to the original V10 version.

Regarding this particular part of code, are you sure that the conditionally invisible fields are necessary? The original tree view doesn't have those two fields and therefore it's open for being customized.

alexey-pelykh replied:
Well, those fields were added for `department_manager` validation policy, so for all others these fields are not valuable to be seen

astirpe replied:
As I said, I think it doesn't worth it.

alexey-pelykh replied:
Does not worth adding `department_manager` policy? Does not worth hiding policy-specific columns?

astirpe replied:
Not by adding such black magic in this module.

I would love to continue to discuss about this, but my holidays are now starting.

alexey-pelykh replied:
:) It's not black magic, it's even used in Odoo itself :P Please ping me when you're back from holidays

astirpe reviewed (changes requested):
I think the most of the proposed changes should be moved to a separate module, as the approval process of a timesheet sheet can vary a lot among different companies. Trying to find a generic way for all the kind of companies seems to me something debatable.

alexey-pelykh commented:
@astirpe this was brought up by @mreficent earlier, the actual validation models can be extracted to a separate module, but the hooking points and related refactor has to be done within this module. In other words, there are approximately 20 lines of code that implement different models, all other are needed to allow those models to act. On top of that - this module is extendable, and I've decided to implement most commonly-used models within it.

alexey-pelykh commented:
In other words, if specific company has group-based validation model or something - it's possible to make an extension to this module that would have approximately 10-20 lines of code in it that actually implement one extra mode of validation.

MiquelRForgeFlow commented:
I would call the new module as "`hr_timesheet_sheet_review_policy`" or "`hr_timesheet_sheet_approval_policy`"

alexey-pelykh commented:
@mreficent to make that module possible, `hr_timesheet_sheet` module still has to be changed, to have concept of customizable validation models. Due to this fact, I don't understand value of ejecting commonly-used policies to separate module

alexey-pelykh commented on hr_timesheet_sheet/models/hr_timesheet_sheet.py:
```diff
@@ -164,6 +199,59 @@ def _default_employee(self):
+        return res
+
+    @api.model
+    def _modify_view_for_policy(self, view_type, view):
```
@mreficent @astirpe to give you better insights: this method has 4 lines of code specific to validation models.

alexey-pelykh commented on hr_timesheet_sheet/models/hr_timesheet_sheet.py:
```diff
@@ -210,28 +336,49 @@ def _check_start_end_dates(self):
 
     @api.multi
-    def _get_overlapping_sheet_domain(self, user_id=False):
+    def _get_complete_name_components(self):
```
@mreficent @astirpe to give you better insights: this method has another 4 lines of code specific to validation models.

alexey-pelykh commented on hr_timesheet_sheet/models/hr_timesheet_sheet.py:
```diff
@@ -210,28 +336,49 @@ def _check_start_end_dates(self):
+        return result
+
+    @api.multi
+    def _get_overlapping_sheet_domain(self):
```
@mreficent @astirpe to give you better insights: this method has another 4 lines of code specific to validation models.

alexey-pelykh commented on hr_timesheet_sheet/models/hr_timesheet_sheet.py:
```diff
@@ -273,6 +430,21 @@ def _check_company_id_add_line_task_id(self):
                       'the Task must be the same.'))
 
+    @api.multi
+    def _get_possible_reviewers(self):
```
@mreficent @astirpe to give you better insights: this method has another 6 lines of code specific to validation models.

alexey-pelykh commented on hr_timesheet_sheet/models/hr_timesheet_sheet.py:
```diff
@@ -498,21 +695,55 @@ def action_timesheet_confirm(self):
+
+    @api.multi
+    def _check_can_review(self):
+        for sheet in self.filtered(lambda x: not x.can_review):
```
@mreficent @astirpe to give you better insights: this method has another 12 lines of code specific to validation models.

alexey-pelykh commented:
@mreficent @astirpe I walked over the code of the PR, there are approx. 30 lines of code that are related to other validation models apart of default one. Do you honestly think it worth creating 3 extra models, one per each validation model, to add approx. 10 lines of code per each module?

alexey-pelykh commented:
@tarteo if you have a spare minute, it would be great to get your review as well

alexey-pelykh commented:
@CasVissers maybe you're interested in reviewing this?

alexey-pelykh commented:
Closed in favor of #272
